File: crates/apollia-desktop/ui/src/lib/ipc/planMode.ts
// Typed IPC wrappers for the plan-mode approval gate.
//
// Mirrors the real runtime payload: the engine emits a `PlanApprovalRequired`
// event carrying the generated plan, and the operator resolves the gate by
// run_id through the `submit_plan_decision` Tauri command.

import { invoke } from "@tauri-apps/api/core";

/** A single step of a proposed plan, as carried by the runtime event. */
export interface PlanStep {
  step_id: string;
  description: string;
  depends_on: string[];
  tool_hint?: string | null;
  model_hint?: string | null;
}

/** Plan awaiting an operator decision, extracted from `PlanApprovalRequired`. */
export interface ProposedPlan {
  run_id: string;
  plan_id: string;
  task_id: string;
  step_count: number;
  steps: PlanStep[];
  ttl_secs: number;
}

/**
 * Operator decision sent back to the gate.
 *
 * Internally tagged on `decision` so it deserializes into the Rust
 * `PlanDecisionDto`. The `edit` variant is wired by the edit flow.
 */
export type PlanDecision =
  | { decision: "approve" }
  | { decision: "reject"; reason?: string }
  | { decision: "edit"; revised_steps: PlanStep[] };

/**
 * Submits the operator's decision for the plan pending approval on `runId`.
 *
 * Resolves to `void` on success; rejects with the runtime error string when no
 * gate is pending for `runId` or the registry is unavailable.
 */
export async function submitPlanDecision(
  runId: string,
  decision: PlanDecision,
): Promise<void> {
  return invoke<void>("submit_plan_decision", { runId, decision });
}

// ── Conversational (session-keyed) plan mode ──────────────────────────────────
//
// Distinct from the run-keyed ORIA gate above: these wrappers target the
// chat-native soft gate, identified by `sessionId`. They back the header toggle
// chip (`setPlanMode`) and the in-conversation review card (`approvePlan` /
// `rejectPlan`).

/**
 * Reads the runtime-level default plan-mode state for new chat sessions.
 *
 * The value comes from the `[chat] plan_mode_default` key of `apollia.toml`,
 * read once at boot. The desktop uses it to seed its per-user preference, so the
 * config stays the single source of truth for the default.
 *
 * Resolves to the boolean default; rejects with the runtime error string when
 * the runtime handle is unavailable.
 */
export async function getPlanModeDefault(): Promise<boolean> {
  return invoke<boolean>("get_plan_mode_default");
}

/**
 * Enables or disables plan mode for a chat session.
 *
 * Resolves to `void` on success; rejects with the runtime error string when the
 * chat subsystem is unavailable or the session does not exist.
 */
export async function setPlanMode(
  sessionId: string,
  enabled: boolean,
): Promise<void> {
  return invoke<void>("set_plan_mode", { sessionId, enabled });
}

/** Snapshot returned by `get_chat_plan`: the plan plus the session phase. */
export interface ChatPlanSnapshot {
  /** Current plan payload (the unified `apollia_core::plan::Plan`) or null. */
  plan: unknown;
  /** Session plan-mode phase (the authoritative gate state). */
  phase: string;
}

/**
 * Reads the current plan snapshot for a session.
 *
 * Lets the plan tab hydrate on mount from authoritative runtime state instead of
 * relying solely on live events, so a plan produced before the tab was opened
 * still renders. The `phase` is the gate state: an executed plan keeps its
 * `awaiting_approval` status on the plan object, but its session phase has moved
 * on, so the caller decides the approval card from `phase`, not the plan status.
 */
export async function getChatPlan(sessionId: string): Promise<ChatPlanSnapshot> {
  return invoke<ChatPlanSnapshot>("get_chat_plan", { sessionId });
}

/**
 * Approves the submitted plan for a session (soft conversational gate).
 *
 * Resolves to `void` on success; rejects with the runtime error string when the
 * session is no longer awaiting approval or the chat subsystem is unavailable.
 */
export async function approvePlan(sessionId: string): Promise<void> {
  return invoke<void>("approve_plan", { sessionId });
}

/**
 * Rejects the submitted plan for a session with an optional reason.
 *
 * Resolves to `void` on success; rejects with the runtime error string when the
 * session is no longer awaiting approval or the chat subsystem is unavailable.
 */
export async function rejectPlan(
  sessionId: string,
  reason?: string,
): Promise<void> {
  return invoke<void>("reject_plan", { sessionId, reason });
}

File: crates/apollia-desktop/src/commands/plan_mode.rs
//! Tauri IPC command for the plan-mode approval gate.
//!
//! When the ORIA engine pauses after plan generation it emits
//! `RuntimeEvent::PlanApprovalRequired` and awaits a decision. This command is
//! the desktop resolver: it forwards the operator's decision (approve or
//! reject) to the pending gate through
//! [`apollia_runtime::plan_approval::PlanApprovalHandle`].

use apollia_core::plan::PlanMutation;
use apollia_runtime::embedded::RuntimeHandle;
use apollia_runtime::plan_approval::PlanApprovalHandle;
use tauri::State;

/// Operator decision on a proposed plan, deserialized from the frontend.
///
/// Internally tagged on the `decision` field so the Svelte layer sends
/// `{ decision: "approve" }` or `{ decision: "reject", reason: "..." }`.
#[derive(Debug, serde::Deserialize)]
#[serde(tag = "decision", rename_all = "snake_case")]
pub enum PlanDecisionDto {
    /// Approve the plan as generated; execution starts.
    Approve,
    /// Reject the plan, with optional feedback injected into replanning.
    Reject {
        /// Free-text guidance for the next planning attempt.
        #[serde(default)]
        reason: Option<String>,
    },
    /// Submit an edited plan, executed directly without replanning.
    Edit {
        /// Revised steps (unified plan step), validated by the engine before
        /// execution.
        revised_steps: Vec<apollia_core::plan::PlanStep>,
    },
}

/// Submits the operator's decision on the plan awaiting approval for `run_id`.
///
/// Resolves the pending gate via [`PlanApprovalHandle`]: approval unblocks
/// execution, rejection triggers replanning with the optional reason.
///
/// # Errors
///
/// Returns `Err(String)` when the plan-gate registry is unavailable, or when no
/// gate is pending for `run_id` (unknown, already resolved, or expired).
#[tauri::command]
pub async fn submit_plan_decision(
    state: State<'_, RuntimeHandle>,
    run_id: String,
    decision: PlanDecisionDto,
) -> Result<(), String> {
    let gates = state
        .plan_gates
        .as_ref()
        .ok_or_else(|| "plan-gate registry not available".to_string())?;
    let handle = PlanApprovalHandle::new(gates.clone());

    match decision {
        PlanDecisionDto::Approve => {
            handle.approve(&run_id).map_err(|e| e.to_string())?;
            tracing::info!(run_id = %run_id, "plan.gate.approved");
        }
        PlanDecisionDto::Reject { reason } => {
            handle.reject(&run_id, reason).map_err(|e| e.to_string())?;
            tracing::info!(run_id = %run_id, "plan.gate.rejected");
        }
        PlanDecisionDto::Edit { revised_steps } => {
            handle
                .edit(&run_id, revised_steps)
                .map_err(|e| e.to_string())?;
            tracing::info!(run_id = %run_id, "plan.gate.edited");
        }
    }

    Ok(())
}

/// Returns the runtime-level default plan-mode state for new chat sessions.
///
/// Read from the `[chat] plan_mode_default` key of `apollia.toml` at boot. The
/// desktop seeds its per-user preference from this value so the config stays the
/// single source of truth for the default.
#[tauri::command]
pub async fn get_plan_mode_default(state: State<'_, RuntimeHandle>) -> Result<bool, String> {
    Ok(state.plan_mode_default)
}

/// Enables or disables plan mode for `session_id`.
///
/// Delegates to the chat manager, which initializes the phase to `Discovery`
/// when enabling and resets it to a neutral state when disabling. Returns an
/// error string when the chat subsystem is unavailable or the session does not
/// exist.
///
/// # Errors
///
/// Returns `Err(String)` when no chat manager is wired into the runtime handle,
/// or when the toggle targets a session that does not exist.
#[tauri::command]
pub async fn set_plan_mode(
    state: State<'_, RuntimeHandle>,
    session_id: String,
    enabled: bool,
) -> Result<(), String> {
    let manager = state
        .chat_manager
        .as_ref()
        .ok_or_else(|| "chat subsystem not available".to_string())?;
    manager
        .set_plan_mode(session_id, enabled)
        .await
        .map_err(|e| e.to_string())
}

/// Approves the plan awaiting approval for `session_id` (soft conversational gate).
///
/// Moves the session to the executing phase, emits `ChatPlanApproved`, and
/// resumes the agent so it executes the approved plan. Returns an error string
/// when the chat subsystem is unavailable or the session is not awaiting
/// approval.
///
/// # Errors
///
/// Returns `Err(String)` when no chat manager is wired into the runtime handle,
/// when the session does not exist, or when it is not awaiting approval.
#[tauri::command]
pub async fn approve_plan(
    state: State<'_, RuntimeHandle>,
    session_id: String,
) -> Result<(), String> {
    let manager = state
        .chat_manager
        .as_ref()
        .ok_or_else(|| "chat subsystem not available".to_string())?;
    manager
        .approve_plan(&session_id)
        .await
        .map_err(|e| e.to_string())
}

/// Rejects the plan awaiting approval for `session_id` with an optional reason.
///
/// Emits `ChatPlanRejected` and triggers a revision turn; the session stays in
/// the awaiting-approval phase (the gate is soft). Returns an error string when
/// the chat subsystem is unavailable or the session is not awaiting approval.
///
/// # Errors
///
/// Returns `Err(String)` when no chat manager is wired into the runtime handle,
/// when the session does not exist, or when it is not awaiting approval.
#[tauri::command]
pub async fn reject_plan(
    state: State<'_, RuntimeHandle>,
    session_id: String,
    reason: Option<String>,
) -> Result<(), String> {
    let manager = state
        .chat_manager
        .as_ref()
        .ok_or_else(|| "chat subsystem not available".to_string())?;
    manager
        .reject_plan(&session_id, reason)
        .await
        .map_err(|e| e.to_string())
}

/// Returns the ordered plan mutation history recorded for `session_id`.
///
/// Mutations come back in insertion order (oldest first), including removal
/// tombstones, so the desktop scrubber can replay the plan construction
/// revision by revision. A known session with no recorded mutations yields an
/// empty list.
///
/// # Errors
///
/// Returns `Err(String)` when the chat subsystem is unavailable, when the
/// session does not exist, or when the plan history read fails.
#[tauri::command]
pub async fn list_plan_mutations(
    state: State<'_, RuntimeHandle>,
    session_id: String,
) -> Result<Vec<PlanMutation>, String> {
    let manager = state
        .chat_manager
        .as_ref()
        .ok_or_else(|| "chat subsystem not available".to_string())?;
    manager
        .read_plan_mutations(session_id)
        .await
        .map_err(|e| e.to_string())
}
